### PR TITLE
TubeGenerator weighed UV mapping

### PR DESCRIPTION
**Pull request by sandord**
Weighed UV mapping support for TubeGenerator.

This will fix UV mapping for non-circular profiles and for non-equally stepped paths as well.

Default (torus) UV mapping (on a mesh that has a flat inside area):

image

Weighed UV mapping applied:

image

You need to set `WeighedUV` to `true` on the `TubeGenerator` to enable this feature.

I also fixed

> float uv_around = (float)j / (float)(~~nRings~~ nRingSize);

for non-weighed UV mapping.

There still a bug from the original code remaining that causes the closing triangles (when `ClosedLoop` is `true`) to connect vertices from the end of the path to vertices from the beginning of the path, resulting in an extremely compressed texture:

image

I'd like to fix this bug as well to complete the UV mapping improvements but until now I haven't been able to get it right. Any help would be greatly appreciated.

### CODE REVIEW

**Comment by sandord** [comment]
Also, I'm not sure if I got the terminology right, I sort of made up the 'weighed UV' thing. If there's a more appropriate name for this, please let me know.

File: mesh_generators/GenCylGenerators.cs [hunk 1]
```diff
@@ -79,38 +80,71 @@ override public MeshGenerator Generate()
...
+            float running_uv_along = 0;
+
             // generate tube
             for (int ri = 0; ri < nRings; ++ri) {
 
                 // propagate frame
                 if (ri != 0) {
                     Vector3d tan = CurveUtils.GetTangent(Vertices, ri);
                     fCur.Origin = (Vector3f)Vertices[ri];
-                    if (ri == 11)
```
Comment (sandord) [inline mesh_generators/GenCylGenerators.cs]: I removed this because it didn't seem to do anything, the `dv` variable isn't used anywhere.

**Comment by sandord** [comment]
I think the artifact shown in the last image is caused by a few triangles that are composed from vertices from both the beginning and the end of the tube.

**Comment by rms80** [comment]
I merged this into the master branch. I rewrote bits of it so that:
1) this is the default and only behavior. I don't see a reason to keep the old behavior.
2) fixed issue w/ loops (start/end vertices were not duplicated)
3) simplified code by using some standard functions

(also fixed various other bugs in this class at the same time)

thanks for your contribution!

**Comment by sandord** [comment]
Great!

By the way, my reasoning behind the feature switch was:

- for those who don't need the new behavior, performance won't be degraded (not sure how significant this is) 
- existing depending projects may rely on the old behavior and they'd have no way of opting out

**Pull request merged**